### ISSUE TITLE
How to deal with variable sized arrays in the observation.

### ISSUE THREAD

**Issue by prakhar6sharma** [aai-institute/tianshou#719]
- [x] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [ ] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [ ] new feature request
- [x] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [x] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, gym, torch, numpy, sys
  print(tianshou.__version__, gym.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
  0.4.9 0.25.0 1.12.0+cu102 1.21.6 3.7.13 (default, Jul 13 2022, 10:47:32)
  [GCC 9.4.0] linux
  ```
  
  I hope you are having a great day!!!
  
  My environment returns a dict as the obs (shown below).
 ```
 obs = {
            'node': self.node_vector_representation,
            'jobs': job_vector_representation,
            'ready_jobs': ready_jobs_vector_representation,
            'node_job_dependency': job_node_dependency_mapping
        }
```
Now across different steps, I have the size of the `jobs`, `ready_jobs`, & `node_job_dependency` changing. For example, in step 1 the shapes of these objects are `[4, 10], [4], [4, 5]` and in the step 2 the shape of these objects change to `[3, 10], [3], [3, 5]` because there is one less job in the system.

Now this problem could have been altogether avoided by keeping the size of jobs fixed (via padding). But as asked in #706 , I want to have support for variable actions (i.e. jobs available to choose at any given point in time is changing).

Running the collector with this type of dict obs results in an error by the `__setitem__` function in `batch.py`. #582 discusses how to solve this issue by changing the dtype of arrays to object but it then results in an error when converting these values to a torch.tensor for forward propagation.

I would like to know if there is any other efficient solution to this problem of batching.

Thanks.

**Comment by Trinkle23897** [comment]
> when converting these values to a torch.tensor for forward propagation.

Maybe change `torch.tensor` to `torch.nested_tensor`? https://pytorch.org/docs/master/nested.html

**Comment by prakhar6sharma** [comment]
Thanks @Trinkle23897 for the suggestion. I have found a work around using this.
